### PR TITLE
LP-1962 Create Generic Multi-Select Choices Fields

### PR DESCRIPTION
**Pull request by qasimgulzar**
### Description

[LP-1962](https://philanthropyu.atlassian.net/browse/LP-1962)

### CODE REVIEW

**Comment by philanthropyu** [comment]
Can one of the admins verify this patch?

File: openedx/custom/form_fields.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                ))
+                if subindex is not None:
+                    subindex += 1
+        return groups
+
+
+class MultiSelectFormFieldWithOthers(forms.MultipleChoiceField):
+    widget = CheckboxSelectMultipleWithOther
+
+    def validate(self, value):
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply openedx/custom/form_fields.py:95]: @saad-waqar not really but if you think there is any case, please mention I will handle that too.

File: openedx/custom/form_fields.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        """Validate that the input is a list or tuple."""
+        if self.required and not value:
+            raise ValidationError(self.error_messages['required'], code='required')
+        # Validate that each value in the value list is in self.choices.
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply openedx/custom/form_fields.py:99]: will remove and resolve.

File: openedx/custom/db/fields.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        '''
+        this function is to validate the input values for multi select field, however we are implementing field with
+        support of other input filed we are disabling validations to let other input text pass to the database.
+        :param value: list of all selected choice with other text.
+        :param model_instance: current model instance for with it is saving data.
+        :return: nothing
+        '''
+        for opt_select in value:
+            if '|' in opt_select:
+                raise exceptions.ValidationError(self.error_messages['invalid_char'] % value)
```
Comment (zeshanarifios) [inline openedx/custom/db/fields.py:54]: This validation is not enough for this model field, We are missing basic required validation. If field is required but user neither select any field nor enter text in other textfield it should raise validation error for required.. We are also not handling max_length validation here.. 
Form has required validation but model should also have it.
Please update this method with proper validations

File: openedx/custom/forms/fields.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            'type': 'text' if value == 'other' else self.input_type,
+            'is_other': label == 'Other'
+        })
+        return option
+
+    def optgroups(self, name, value, attrs=None):
+        """Return a list of optgroups for this widget."""
+
+        OTHER_VALUE_INDEX_ALWAYS = 0
+        other_values = [val for val in value if not any(val in choice for choice in self.choices)]
```
Comment (zeshanarifios) [inline openedx/custom/forms/fields.py:23]: Please update it to 
```
choice_values = [choice[0] for choice in self.choices]
other_values = [val for val in value if not val in choice_values]
```

File: openedx/custom/forms/fields.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            'value': self.other_choice if value == 'other' else value,
+            'type': 'text' if value == 'other' else self.input_type,
+            'is_other': label == 'Other'
+        })
+        return option
+
+    def optgroups(self, name, value, attrs=None):
+        """Return a list of optgroups for this widget."""
+
+        OTHER_VALUE_INDEX_ALWAYS = 0
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply openedx/custom/forms/fields.py:22]: it is only used in current scope

**Comment by zeshanarifios** [comment]
@qasimgulzar Please also run isort on all files where you add imports if you haven't run already.

File: openedx/custom/forms/fields.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/custom/forms/fields.py:22]: Let's please do other things except moving to constants

File: openedx/custom/db/fields.py (hunk 3 above)
Comment (qasimgulzar) [reply openedx/custom/db/fields.py:54]:
```
class MultiSelectField(models.CharField):
    """ Choice values can not contain commas. """

    def __init__(self, *args, **kwargs):
        self.min_choices = kwargs.pop('min_choices', None)
        self.max_choices = kwargs.pop('max_choices', None)
        super(MultiSelectField, self).__init__(*args, **kwargs)
        self.max_length = get_max_length(self.choices, self.max_length)
        self.validators[0] = MaxValueMultiFieldValidator(self.max_length)
        if self.min_choices is not None:
            self.validators.append(MinChoicesValidator(self.min_choices))
        if self.max_choices is not None:
            self.validators.append(MaxChoicesValidator(self.max_choices))
```

it is already handled in parent constructor with min and max choices.

Comment (zeshanarifios) [reply openedx/custom/db/fields.py]: Those are different validations brother, Those are about minimum number of choices and max number of choices. 
 
I am talking about **required** and **max_length** validation.

File: openedx/custom/forms/fields.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        other_values = [val for val in value if not any(val in choice for choice in self.choices)]
+        other_values = '' if not other_values else other_values.pop(OTHER_VALUE_INDEX_ALWAYS)
+
+        self.other_choice = other_values
```
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/custom/forms/fields.py:28]: Please have a look, It is missed

File: openedx/custom/forms/fields.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+        groups = super(CheckboxSelectMultipleWithOther, self).optgroups(name, value, attrs)
+        return groups
```
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/custom/forms/fields.py:31]: Please have a look, It is missed

File: openedx/custom/db/fields.py (hunk 3 above)
Comment (qasimgulzar) [reply openedx/custom/db/fields.py]: @zeshanarifios both are already implemented, as you have also verified on your machine.

**Pull request closed without merging**